### PR TITLE
REFACTOR: Extract and move FeeBudget from xgap

### PR DESCRIPTION
**Pull request by narumiruna**

### CODE REVIEW

**Review by kbearXD — APPROVED** [review]

File: pkg/strategy/common/fee_budget.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+type State struct {
+	AccumulatedFeeStartedAt time.Time                   `json:"accumulatedFeeStartedAt,omitempty"`
+	AccumulatedFees         map[string]fixedpoint.Value `json:"accumulatedFees,omitempty"`
+	AccumulatedVolume       fixedpoint.Value            `json:"accumulatedVolume,omitempty"`
+}
+
+func (s *State) IsOver24Hours() bool {
+	return time.Since(s.AccumulatedFeeStartedAt) >= 24*time.Hour
```
Comment (c9s) [inline pkg/strategy/common/fee_budget.go:78]: can we move this to util? it seems like we use it a lot

File: pkg/strategy/common/fee_budget.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+	f.State.AccumulatedFees[trade.FeeCurrency] = f.State.AccumulatedFees[trade.FeeCurrency].Add(trade.Fee)
+	f.State.AccumulatedVolume = f.State.AccumulatedVolume.Add(trade.Quantity)
+	log.Infof("[FeeBudget] accumulated fee: %s %s", f.State.AccumulatedFees[trade.FeeCurrency].String(), trade.FeeCurrency)
+}
+
+type State struct {
+	AccumulatedFeeStartedAt time.Time                   `json:"accumulatedFeeStartedAt,omitempty"`
+	AccumulatedFees         map[string]fixedpoint.Value `json:"accumulatedFees,omitempty"`
+	AccumulatedVolume       fixedpoint.Value            `json:"accumulatedVolume,omitempty"`
```
Comment (c9s) [inline pkg/strategy/common/fee_budget.go:74]: this volume is unrelated to the fee budget?

File: pkg/strategy/common/fee_budget.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+func (f *FeeBudget) HandleTradeUpdate(trade types.Trade) {
+	log.Infof("[FeeBudget] received trade %s", trade.String())
+
+	if f.State.IsOver24Hours() {
+		f.State.Reset()
+	}
+
+	// safe check
+	if f.State.AccumulatedFees == nil {
+		f.State.AccumulatedFees = make(map[string]fixedpoint.Value)
```
Comment (c9s) [inline pkg/strategy/common/fee_budget.go:63]: you might need a mutex to protect it

**Review by c9s — APPROVED** [review]

File: pkg/strategy/common/fee_budget.go (hunk 2 above)
Comment (narumiruna) [reply pkg/strategy/common/fee_budget.go:74]: oops

File: pkg/strategy/common/fee_budget.go (hunk 3 above)
Comment (narumiruna) [reply pkg/strategy/common/fee_budget.go:63]: ok
